spec/lang/step/terminators.md

```diff
@@ -174,6 +174,25 @@ impl<M: Memory> Machine<M> {
         })
     }
 
+    /// A helper function to evaluate the return expression and prepare it for in-place passing.
+    fn eval_return_place(&mut self, ret_expr: PlaceExpr) -> NdResult<(Place<M>, Type)> {
+        // First evaluate the return place. (Left-to-right!)
+        let (caller_ret_place, caller_ret_ty) = self.eval_place(ret_expr)?;
+        // FIXME: should we care about `caller_ret_place.align`?
+        // Make sure we can use it in-place.
+        self.prepare_for_inplace_passing(caller_ret_place, caller_ret_ty)?;
+        ret((caller_ret_place, caller_ret_ty))
+    }
+
+    /// A helper function to evaluate a function pointer.
+    fn eval_function_ptr(&mut self, fn_ptr: ValueExpr) -> Result<Function> {
```

Comment:
Instead of adding a new function, I'd suggest changing `fn_from_ptr` to take a `Value` rather than a `ThinPointer`. Then this here becomes so simple that it can be inlined where it is needed.